theano/compile/compilelock.py

```diff
@@ -0,0 +1,67 @@
+"""
+Locking mechanism to ensure no two compilations occur simultaneously
+in the same compilation directory (which can cause crashes).
+"""
+import os
+import typing
+from contextlib import contextmanager
+
+import filelock
+
+from theano import config
+
+
+__all__ = [
+    "force_unlock",
+    "lock_ctx",
+]
+
+
+_locks = {}
```

Comment:
The main reason for changing the lock mechanism is to make it work across processes.  It looks like this approach requires shared memory, which is the same shortcoming as the current approach.

What I think we need is a [`fcntl`](https://docs.python.org/3/library/fcntl.html)-like approach that works for all filesystems.

Reply:
Correct me if I'm wrong, but [`filelock`](https://github.com/benediktschmitt/py-filelock) does NOT require shared memory.

Quoting the readme:
> Use the FileLock if all instances of your application are running on the same host and a SoftFileLock otherwise.

The dict in line 20 is just to save time on the creation of the `FileLock` object. The state of the lock is based just on the file on disk.

Reply:
Yes, that looks correct.